enterprise/backend/src/metabase_enterprise/remote_sync/api.clj

```diff
@@ -0,0 +1,279 @@
+   [metabase-enterprise.remote-sync.impl :as impl]
+   [metabase-enterprise.remote-sync.models.remote-sync-object :as remote-sync.object]
+   [metabase-enterprise.remote-sync.models.remote-sync-task :as remote-sync.task]
+   [metabase-enterprise.remote-sync.settings :as settings]
+   [metabase-enterprise.remote-sync.source :as source]
+   [metabase-enterprise.remote-sync.source.ingestable :as source.ingestable]
+   [metabase-enterprise.remote-sync.source.protocol :as source.p]
+   [metabase.api.common :as api]
+   [metabase.api.macros :as api.macros]
+   [metabase.api.routes.common :refer [+auth]]
+   [metabase.app-db.cluster-lock :as cluster-lock]
+   [metabase.collections.models.collection :as collection]
+   [metabase.util.jvm :as u.jvm]
+   [metabase.util.log :as log]
+   [metabase.util.malli.schema :as ms]))
+
+(set! *warn-on-reflection* true)
+
+(defn- run-async!
+  [ttype branch f]
+  (let [{task-id :id
+         existing? :existing?}
+        (cluster-lock/with-cluster-lock impl/cluster-lock
+          (if-let [{id :id} (remote-sync.task/current-task)]
+            {:existing? true :id id}
+            (remote-sync.task/create-sync-task! ttype api/*current-user-id*)))]
+    (api/check-400 (not existing?) "Remote sync in progress")
+    (u.jvm/in-virtual-thread*
+     (dh/with-timeout {:interrupt? true
+                       :timeout-ms (settings/remote-sync-task-time-limit-ms)}
+       (let [result (f task-id)]
+         (case (:status result)
+           :success (do (remote-sync.task/complete-sync-task! task-id)
+                        (settings/remote-sync-branch! branch))
+           :error (remote-sync.task/fail-sync-task! task-id (:message result))
+           (remote-sync.task/fail-sync-task! task-id "Unexpected Error")))))
+    task-id))
+
+(defn- async-import!
+  [branch force? import-args]
+  (let [ingestable-source     (source.p/->ingestable (source/source-from-settings branch) {:path-filters [#"collections/.*"]})
+        source-version        (source.ingestable/ingestable-version ingestable-source)
+        last-imported-version (remote-sync.task/last-import-version)
+        has-dirty?            (remote-sync.object/dirty-global?)]
+    (when (and has-dirty? (not force?))
+      (throw (ex-info "There are unsaved changes in the Remote Sync collection which will be overwritten by the import. Force the import to discard these changes."
+                      {:status-code 400
+                       :conflicts   true})))
+    (if (and (not force?) (= last-imported-version source-version))
+      (do (log/infof "Skipping import: source version %s matches last imported version" source-version)
+          (settings/remote-sync-branch! branch)
+          nil)
+      (run-async! "import" branch (fn [task-id] (impl/import! ingestable-source task-id import-args))))))
+
+(defn- async-export!
+  [branch message]
+  (let [source (source/source-from-settings branch)
+        last-task-version (remote-sync.task/last-version)
```

Comment:
Should this be a concern of the impl instead of happening here? I would like to expose fewer implementation details of the `source` module here. That means the API just concered with the plumping of running the task and the impl handles all the lower level details.